File: src/lib/ipc/fs.ts
import { invoke } from '@tauri-apps/api/core';
import { listen, type UnlistenFn } from '@tauri-apps/api/event';
import { platform } from './rpc';

export interface FsEntry {
  name:     string;
  path:     string;
  is_dir:   boolean;
  size:     number | null;
  modified: number | null;  // Unix timestamp ms
  created:  number | null;  // Unix timestamp ms (null when the FS has no birth time)
}

export interface FsRoot {
  name: string;
  path: string;
  kind: 'home' | 'desktop' | 'documents' | 'downloads' | 'drive' | 'wsl';
}

/** Read a directory — returns entries with metadata. Dot-prefixed entries
 *  are skipped unless `showHidden` is set. */
export const fsReadDir = (path: string, showHidden = false) =>
  platform<FsEntry[]>('fs_read_dir', { path, show_hidden: showHidden });

/** Return quick-access roots (common dirs + drives). */
export const listFsRoots = () =>
  platform<FsRoot[]>('list_fs_roots');

/** Installed WSL distributions as `\\wsl.localhost\<distro>` roots (Windows;
 *  empty elsewhere or when WSL isn't installed). */
export const listWslDistros = () =>
  platform<FsRoot[]>('list_wsl_distros');

export const fsCreateDir      = (path: string)                      => platform<void>('fs_create_dir',        { path });
export const fsCreateFile     = (path: string)                      => platform<void>('fs_create_file',       { path });
export const fsWriteTextFile  = (path: string, content: string)     => platform<void>('fs_write_text_file',   { path, content });
export const fsReadTextFile   = (path: string)                      => platform<string>('fs_read_text_file', { path });
export const fsRename         = (oldPath: string, newPath: string)  => platform<void>('fs_rename',            { old_path: oldPath, new_path: newPath });
export const fsDelete         = (path: string)                      => platform<void>('fs_delete',            { path });

// ── File explorer: copy / move / delete / open / watch ─────────────────────
/** Copy entries into `destDir`; returns the created destination paths. With
 *  `overwrite`, same-named items merge into / replace the existing entry
 *  instead of getting a " (2)" suffix. */
export const fsCopy        = (sources: string[], destDir: string, overwrite = false, opId?: string) => invoke<string[]>('fs_copy', { sources, destDir, overwrite, opId: opId ?? null });
/** Move (cut+paste) entries into `destDir`; returns the new paths. With
 *  `overwrite`, same-named items merge into / replace the existing entry. */
export const fsMove        = (sources: string[], destDir: string, overwrite = false, opId?: string) => invoke<string[]>('fs_move', { sources, destDir, overwrite, opId: opId ?? null });
/** Duplicate entries in place ("file (2).ext"). Returns the created paths. */
export const fsDuplicate   = (paths: string[], opId?: string) => invoke<string[]>('fs_duplicate', { paths, opId: opId ?? null });
/** Request cancellation of a running copy/move/duplicate by its op id. */
export const fsCancelOp    = (opId: string) => invoke<void>('fs_cancel_op', { opId });
/** One old→new rename pair for a batch rename. */
export interface RenamePair { from: string; to: string; }
/** Batch-rename many entries atomically (two-phase, collision-safe). */
export const fsRenameMany  = (pairs: RenamePair[]) => platform<string[]>('fs_rename_many', { pairs });

/** Recursive size of a folder (bytes + file/dir counts). */
export interface DirSize { bytes: number; files: number; dirs: number; }
export const fsDirSize     = (path: string) => platform<DirSize>('fs_dir_size', { path });
/** Combined recursive size of several paths (multi-selection footer). */
export const fsPathsSize   = (paths: string[]) => platform<DirSize>('fs_paths_size', { paths });

/** Per-drive storage usage for the Overview dashboard. */
export interface DriveUsage { name: string; path: string; total: number | null; free: number | null; }
export interface OverviewStats { drives: DriveUsage[]; total_capacity: number; total_free: number; }
export const fsOverviewStats = () => platform<OverviewStats>('fs_overview_stats');

/** One item in the Recycle Bin / trash. */
export interface TrashEntry { id: string; name: string; original_path: string; deleted_at: number | null; }
/** List trash items (newest first). Empty on macOS. */
export const fsTrashList    = () => platform<TrashEntry[]>('fs_trash_list');
/** Restore trashed items to their original location (Windows/Linux). */
export const fsTrashRestore = (ids: string[]) => platform<void>('fs_trash_restore', { ids });
/** Permanently delete trashed items (Windows/Linux). */
export const fsTrashPurge   = (ids: string[]) => platform<void>('fs_trash_purge', { ids });
/** Empty the whole Recycle Bin (Windows/Linux). */
export const fsTrashEmpty   = () => platform<void>('fs_trash_empty');

/** Progress event for a long-running file operation (copy/move/duplicate). */
export interface FsOpProgress {
  op_id: string;
  kind: 'copy' | 'move' | 'duplicate';
  done_files: number;
  total_files: number;
  done_bytes: number;
  total_bytes: number;
  current: string;
}
/** Subscribe to this window's file-operation progress events. */
export const onFsOpProgress = (cb: (p: FsOpProgress) => void): Promise<UnlistenFn> =>
  listen<FsOpProgress>('arbor://fs-op-progress', e => cb(e.payload));
/** Move entries to the OS trash / Recycle Bin (recoverable). */
export const fsTrash       = (paths: string[]) => platform<void>('fs_trash', { paths });
/** Restore previously-trashed entries to their original locations (undo of
 *  `fsTrash`). Windows / Linux only. */
export const fsUntrash     = (paths: string[]) => platform<void>('fs_untrash', { paths });
/** Permanently delete entries from disk (Shift+Delete). */
export const fsDeleteMany  = (paths: string[]) => platform<void>('fs_delete_many', { paths });
/** Recursively search `root` for entries whose name matches `query` (glob when
 *  it contains `*`/`?`, else case-insensitive substring). Capped at `limit`
 *  (default 5000). Each result carries its full path. */
export const fsSearch = (root: string, query: string, showHidden = false, limit?: number) =>
  platform<FsEntry[]>('fs_search', { root, query, show_hidden: showHidden, limit });
/** Compress `sources` into a new ZIP named `archiveName` inside `destDir`
 *  (collision-resolved). Returns the created archive path. */
export const fsZip = (sources: string[], destDir: string, archiveName: string) =>
  platform<string>('fs_zip', { sources, dest_dir: destDir, archive_name: archiveName });
/** Extract a ZIP `archive`, into `destDir` or — when omitted — a new sibling
 *  folder named after the archive. Returns the destination folder path. */
export const fsUnzip = (archive: string, destDir?: string) =>
  platform<string>('fs_unzip', destDir ? { archive, dest_dir: destDir } : { archive });
/** Set an image file as the desktop wallpaper (Windows / macOS / GNOME). */
export const fsSetWallpaper = (path: string) => invoke<void>('fs_set_wallpaper', { path });
/** Open a path with the OS default app (file) or file manager (dir). */
export const fsOpenDefault = (path: string) => invoke<void>('fs_open_default', { path });
/** Reveal a path in the OS file manager, selecting it. */
export const fsRevealInDir = (path: string) => invoke<void>('fs_reveal_in_dir', { path });
/** Open the OS terminal rooted at `path` (the folder, or a file's parent),
 *  detached so it outlives Arbor. Windows Terminal / cmd · Terminal.app ·
 *  the first available Linux terminal emulator. */
export const fsOpenTerminal = (path: string) => invoke<void>('fs_open_terminal', { path });
/** Expand `%VAR%` / `$VAR` / leading `~` in a typed path. The virtual names
 *  `appdata` / `localappdata` / `home` resolve cross-platform, so `%appdata%`
 *  works on every OS. Returns the input unchanged when there's nothing to expand. */
export const fsExpandPath = (path: string) => platform<string>('fs_expand_path', { path });
/** Open the built-in explorer window at a path (focusing/reusing it per the
 *  one-window setting). `reveal = true` selects the file inside its folder;
 *  `reveal = false` just opens the folder. Used when the user routes the app's
 *  "Open / Reveal in File Explorer" actions to the built-in explorer. */
export const revealInExplorerWindow = (path: string, reveal: boolean) =>
  invoke<void>('reveal_in_explorer', { path, reveal });
/** A pending reveal handed to a freshly-opened explorer window. */
export interface ExplorerRevealPayload { dir: string; select: string | null; }
/** Drain the pending reveal for a window label (explorer window, on mount). */
export const takeExplorerReveal = (label: string) =>
  invoke<ExplorerRevealPayload | null>('take_explorer_reveal', { label });

// ── Cross-window clipboard (copy / cut / paste between explorer windows) ─────
export type ClipOp = 'copy' | 'cut';
/** The shared explorer clipboard payload (process-wide, mirrored per window). */
export interface ClipData { op: ClipOp; paths: string[]; }
/** Set the shared clipboard; broadcasts `arbor://explorer-clip-changed`. */
export const explorerClipSet = (op: ClipOp, paths: string[]) =>
  invoke<void>('explorer_clip_set', { op, paths });
/** Read the shared clipboard (seed a window's local mirror on mount). */
export const explorerClipGet = () => invoke<ClipData | null>('explorer_clip_get');
/** Clear the shared clipboard (after a cut→paste move); broadcasts the change. */
export const explorerClipClear = () => invoke<void>('explorer_clip_clear');

// ── Cross-window drag & drop (overlay ghost + drop hit-testing) ──────────────
/** Ensure the shared drag-ghost overlay window exists (built once, reused). */
export const ensureDragOverlay = () => invoke<void>('ensure_drag_overlay');
/** Show the overlay with `text` at logical screen coordinates `x`/`y`. */
export const dragOverlayShow = (text: string, x: number, y: number) =>
  invoke<void>('drag_overlay_show', { text, x, y });
/** Move the overlay to logical screen coordinates `x`/`y`. */
export const dragOverlayMove = (x: number, y: number) =>
  invoke<void>('drag_overlay_move', { x, y });
/** Hide the overlay (drag ended, or cursor re-entered the source window). */
export const dragOverlayHide = () => invoke<void>('drag_overlay_hide');
/** Drain the current overlay label (overlay window pulls this on mount). */
export const getDragOverlayText = () => invoke<string>('get_drag_overlay_text');
/** On drop, hand the dragged paths to another explorer window under the cursor
 *  (logical screen coords). Returns true when a target window was notified. */
export const explorerDropDispatch = (sourceLabel: string, x: number, y: number, paths: string[]) =>
  invoke<boolean>('explorer_drop_dispatch', { sourceLabel, x, y, paths });
/** Open the OS-native Properties dialog for a path (Windows property sheet /
 *  macOS Finder Get Info / Linux FileManager1 D-Bus). */
export const fsShowProperties = (path: string) => invoke<void>('fs_show_properties', { path });
/** Native system icon for a query (a file extension like ".rs", or an absolute
 *  path — `.exe` yields its embedded icon), as a `data:image/png;base64,…` URI. */
export const fsIcon = (query: string, size: number) => invoke<string>('fs_icon', { query, size });
/** Start watching `path` for changes (replaces any prior watch). Emits the
 *  `arbor://fs-changed` Tauri event on any change in the directory. Pass
 *  `recursive = true` to also catch changes in sub-folders (e.g. a project tree);
 *  the flat explorer leaves it false. */
export const fsWatchStart  = (path: string, recursive = false) =>
  invoke<void>('fs_watch_start', { path, recursive });
/** Stop the active filesystem watch. */
export const fsWatchStop   = () => invoke<void>('fs_watch_stop');
/** Subscribe to this window's `arbor://fs-changed` signal (fired when the watched
 *  directory changes; carries no payload — re-read what you care about). */
export const onFsChanged   = (cb: () => void): Promise<UnlistenFn> =>
  listen('arbor://fs-changed', () => cb());

// ── File explorer: git awareness (TortoiseGit-style overlays + actions) ─────
/** Overlay badge for one explorer entry (or a rolled-up folder). */
export type GitBadge =
  | 'conflicted' | 'modified' | 'deleted' | 'renamed' | 'added' | 'untracked' | 'ignored';

/** Lightweight marker for a child folder that is itself a git repo root. */
export interface GitRepoMarker {
  branch: string | null;
  detached: boolean;
}

export interface FsGitStatus {
  in_repo: boolean;
  repo_root: string | null;
  branch: string | null;
  detached: boolean;
  ahead: number;
  behind: number;
  /** Map keyed by `normPath(entry.path)` → badge. Only non-clean entries present. */
  badges: Record<string, GitBadge>;
  /** Map keyed by `normPath(entry.path)` → marker, for immediate child folders
   *  that are themselves git repo roots (flagged even when `dir` isn't a repo). */
  repos: Record<string, GitRepoMarker>;
}

/** Git status for `dir`'s entries (badges + branch / ahead-behind). Cached per
 *  repo-root; pass `refresh = true` (off the fs watcher) to recompute. */
export const fsGitStatus = (dir: string, refresh = false) =>
  invoke<FsGitStatus>('fs_git_status', { dir, refresh });
/** Stage paths (files / folders / deletions) in their enclosing repo. */
export const fsGitStage   = (paths: string[]) => invoke<void>('fs_git_stage',   { paths });
/** Unstage paths (reset to HEAD). */
export const fsGitUnstage = (paths: string[]) => invoke<void>('fs_git_unstage', { paths });
/** Discard working-tree changes for paths (snapshots to Recovery first). */
export const fsGitDiscard = (paths: string[]) => invoke<void>('fs_git_discard', { paths });
/** Append paths to the repo's `.gitignore` (anchored, folders get a trailing slash). */
export const fsGitIgnore  = (paths: string[]) => invoke<void>('fs_git_ignore',  { paths });
/** Bring the main Arbor window forward and open the repo containing `path`
 *  (delegates the heavy git operations — diff / log / blame — to Arbor's UI). */
export const fsOpenInArbor = (path: string) => invoke<void>('fs_open_in_arbor', { path });

/** One changed file in the staged or unstaged list. */
export interface GitChange {
  /** Absolute path (native separators) — matches `fsReadDir` entry paths. */
  path: string;
  /** Repo-relative path (forward slashes) for display. */
  rel: string;
  badge: GitBadge;
}
/** Working-tree change list for the repo enclosing `dir`. A file staged then
 *  edited again appears in BOTH lists, like `git status`. */
export interface GitChanges {
  repo_root: string | null;
  branch: string | null;
  staged: GitChange[];
  unstaged: GitChange[];
}
/** Full staged/unstaged change list for the repo enclosing `dir`. */
export const fsGitChanges = (dir: string) => invoke<GitChanges>('fs_git_changes', { dir });

/** One local branch of a repo. */
export interface FsBranch { name: string; is_head: boolean; }
/** Local branches of the repo enclosing `path` (sorted, case-insensitive). */
export const fsGitBranches = (path: string) => invoke<FsBranch[]>('fs_git_branches', { path });
/** Switch the repo enclosing `path` to `branch` (safe checkout — fails on
 *  conflicting uncommitted changes). */
export const fsGitCheckout = (path: string, branch: string) => invoke<void>('fs_git_checkout', { path, branch });
/** Remote URL (origin, else first remote) of the repo enclosing `path`, or
 *  `null` when there's no repo / no remote. Used to build "Copy project link". */
export const fsGitRemoteUrl = (path: string) => invoke<string | null>('fs_git_remote_url', { path });

File: src-tauri/src/ipc/platform/fs.rs
//! `fs` domain — pure filesystem operations routed through the platform broker.
//!
//! Each handler is the body the matching `#[tauri::command]` ran inline; the FS
//! logic itself lives in the reusable, Tauri-free [`arbor_fs`] crate
//! (read/write/copy/move/delete/list/search/trash/zip, roots, sizes, path
//! expansion). Handlers map `FsError` → [`AppError`] and otherwise delegate
//! straight to it.
//!
//! ## No more per-command `spawn_blocking`
//!
//! The old async commands wrapped each blocking call in
//! `tokio::task::spawn_blocking`. The generic `rpc` command now runs *every*
//! platform/corvus handler on the blocking pool already (see
//! `crate::commands::rpc_commands::rpc`), so handlers are plain sync functions
//! that call the blocking `arbor-fs` functions directly — off the main thread,
//! identical behavior, minus the wrapper.
//!
//! ## Copy / move / duplicate / cancel — progress through the event sink
//!
//! These four ops stream a throttled `arbor://fs-op-progress` event and own a
//! process-global op-id → [`CancelToken`] registry. A handler reached through
//! the generic `rpc` command holds only `&AppState` (no `AppHandle`), so the
//! progress sink forwards through the **event sink**
//! (`Arc<dyn EventSink>` — [`AppState::event_sink`]) instead of an `AppHandle`.
//! The cancel-token registry is a `OnceLock` static, so it stays reachable from
//! the handler unchanged. Behavior is byte-identical: same topic, same payload
//! shape, same throttling, same registry effects — only the egress handle
//! changed. The old per-command `spawn_blocking` is gone too (see below): the op
//! body runs synchronously on the rpc blocking pool.
//!
//! ## What stays in the shell (NOT here)
//!
//! The OS / Tauri shell-integration that is *not* pure FS I/O remains inline in
//! [`crate::commands::fs_commands`]: open-with-default, reveal-in-file-manager,
//! open-terminal, the native Properties dialog, set-wallpaper, native icons, and
//! the per-window watcher.
//!
//! The pure-FS handlers never touched `AppState`, but the handler macro requires
//! a context first arg, so they take `_state: &AppState` and ignore it. The
//! copy/move/duplicate handlers use it for `event_sink()`.
//!
//! No hooks fire in this domain.

use std::collections::HashMap;
use std::sync::{Arc, Mutex, OnceLock};
use std::time::Instant;

use arbor_ipc::prelude::EventSink;

use crate::error::AppError;
use crate::ipc::platform;
use crate::AppState;
use arbor_fs::prelude::{
    copy, mutate, pathexpand, read, roots, size, trash, zip, CancelToken, DirSize, FsEntry, FsRoot,
    NoopSink, OverviewStats, ProgressSink, RenamePair, TrashEntry,
};

// ---------------------------------------------------------------------------
// Mutating filesystem operations
// ---------------------------------------------------------------------------

#[platform::handler(program = "platform")]
fn fs_create_dir(_state: &AppState, path: String) -> Result<(), AppError> {
    Ok(mutate::create_dir(&path)?)
}

#[platform::handler(program = "platform")]
fn fs_create_file(_state: &AppState, path: String) -> Result<(), AppError> {
    Ok(mutate::create_file(&path)?)
}

#[platform::handler(program = "platform")]
fn fs_rename(_state: &AppState, old_path: String, new_path: String) -> Result<(), AppError> {
    Ok(mutate::rename(&old_path, &new_path)?)
}

#[platform::handler(program = "platform")]
fn fs_rename_many(_state: &AppState, pairs: Vec<RenamePair>) -> Result<Vec<String>, AppError> {
    Ok(mutate::rename_many(&pairs)?)
}

#[platform::handler(program = "platform")]
fn fs_write_text_file(_state: &AppState, path: String, content: String) -> Result<(), AppError> {
    Ok(mutate::write_text(&path, &content)?)
}

#[platform::handler(program = "platform")]
fn fs_read_text_file(_state: &AppState, path: String) -> Result<String, AppError> {
    Ok(read::read_text(&path)?)
}

#[platform::handler(program = "platform")]
fn fs_delete(_state: &AppState, path: String) -> Result<(), AppError> {
    Ok(mutate::delete(&path)?)
}

#[platform::handler(program = "platform")]
fn fs_delete_many(_state: &AppState, paths: Vec<String>) -> Result<(), AppError> {
    Ok(mutate::delete_many(&paths)?)
}

// ---------------------------------------------------------------------------
// Listing + search
// ---------------------------------------------------------------------------

#[platform::handler(program = "platform")]
fn fs_read_dir(
    _state: &AppState,
    path: String,
    show_hidden: Option<bool>,
) -> Result<Vec<FsEntry>, AppError> {
    Ok(read::read_dir(&path, show_hidden.unwrap_or(false))?)
}

#[platform::handler(program = "platform")]
fn fs_search(
    _state: &AppState,
    root: String,
    query: String,
    show_hidden: Option<bool>,
    limit: Option<usize>,
) -> Result<Vec<FsEntry>, AppError> {
    let show_hidden = show_hidden.unwrap_or(false);
    let cap = limit.unwrap_or(5000);
    Ok(read::search(&root, &query, show_hidden, cap)?)
}

// ---------------------------------------------------------------------------
// Trash — recoverable delete + Recycle Bin view
// ---------------------------------------------------------------------------

#[platform::handler(program = "platform")]
fn fs_trash(_state: &AppState, paths: Vec<String>) -> Result<(), AppError> {
    Ok(trash::trash(&paths)?)
}

#[platform::handler(program = "platform")]
fn fs_untrash(_state: &AppState, paths: Vec<String>) -> Result<(), AppError> {
    Ok(trash::untrash(&paths)?)
}

#[platform::handler(program = "platform")]
fn fs_trash_list(_state: &AppState) -> Result<Vec<TrashEntry>, AppError> {
    Ok(trash::trash_list()?)
}

#[platform::handler(program = "platform")]
fn fs_trash_restore(_state: &AppState, ids: Vec<String>) -> Result<(), AppError> {
    Ok(trash::trash_restore(&ids)?)
}

#[platform::handler(program = "platform")]
fn fs_trash_purge(_state: &AppState, ids: Vec<String>) -> Result<(), AppError> {
    Ok(trash::trash_purge(&ids)?)
}

#[platform::handler(program = "platform")]
fn fs_trash_empty(_state: &AppState) -> Result<(), AppError> {
    Ok(trash::trash_empty()?)
}

// ---------------------------------------------------------------------------
// Compress / extract — ZIP archives
// ---------------------------------------------------------------------------

#[platform::handler(program = "platform")]
fn fs_zip(
    _state: &AppState,
    sources: Vec<String>,
    dest_dir: String,
    archive_name: String,
) -> Result<String, AppError> {
    Ok(zip::zip(&sources, &dest_dir, &archive_name)?)
}

#[platform::handler(program = "platform")]
fn fs_unzip(
    _state: &AppState,
    archive: String,
    dest_dir: Option<String>,
) -> Result<String, AppError> {
    Ok(zip::unzip(&archive, dest_dir)?)
}

// ---------------------------------------------------------------------------
// Recursive sizes + quick-access roots + WSL + Overview stats
// ---------------------------------------------------------------------------

#[platform::handler(program = "platform")]
fn fs_dir_size(_state: &AppState, path: String) -> Result<DirSize, AppError> {
    Ok(size::dir_size(&path))
}

#[platform::handler(program = "platform")]
fn fs_paths_size(_state: &AppState, paths: Vec<String>) -> Result<DirSize, AppError> {
    Ok(size::paths_size(&paths))
}

/// Return filesystem quick-access roots (common user dirs + drives / `/`).
#[platform::handler(program = "platform")]
fn list_fs_roots(_state: &AppState) -> Result<Vec<FsRoot>, AppError> {
    Ok(roots::list_roots())
}

/// List installed WSL distributions as navigable roots.
#[platform::handler(program = "platform")]
fn list_wsl_distros(_state: &AppState) -> Result<Vec<FsRoot>, AppError> {
    Ok(roots::list_wsl_distros())
}

/// Real Overview dashboard stats: capacity / free space per drive.
#[platform::handler(program = "platform")]
fn fs_overview_stats(_state: &AppState) -> Result<OverviewStats, AppError> {
    Ok(roots::overview_stats())
}

/// Expand environment-variable references and a leading `~` in a user-typed
/// path (`%appdata%`, `$HOME`, `~/code`, …). Unknown variables are left intact.
#[platform::handler(program = "platform")]
fn fs_expand_path(_state: &AppState, path: String) -> Result<String, AppError> {
    Ok(pathexpand::expand_path(&path))
}

// ---------------------------------------------------------------------------
// Copy / move / duplicate — progress + cooperative cancellation
// ---------------------------------------------------------------------------
// A per-`op_id` cancel token lets the UI abort a running op; a throttled
// `arbor://fs-op-progress` event drives the explorer's progress bar. The pure
// copy/move/duplicate live in `arbor-fs`; here we own the registry and the sink.

/// Process-wide registry of cancel tokens keyed by op_id.
fn cancel_registry() -> &'static Mutex<HashMap<String, CancelToken>> {
    static REG: OnceLock<Mutex<HashMap<String, CancelToken>>> = OnceLock::new();
    REG.get_or_init(|| Mutex::new(HashMap::new()))
}

fn register_op(op_id: &str) -> CancelToken {
    let token = CancelToken::new();
    if let Ok(mut m) = cancel_registry().lock() {
        m.insert(op_id.to_string(), token.clone());
    }
    token
}

fn unregister_op(op_id: &str) {
    if let Ok(mut m) = cancel_registry().lock() {
        m.remove(op_id);
    }
}

/// Request cancellation of a running file operation. No-op for unknown ids
/// (the op may have already finished). Cooperative: the op stops at the next
/// file boundary. Signals the process-global registry only — emits nothing.
#[platform::handler(program = "platform")]
fn fs_cancel_op(_state: &AppState, op_id: String) -> Result<(), AppError> {
    if let Ok(m) = cancel_registry().lock() {
        if let Some(token) = m.get(&op_id) {
            token.cancel();
        }
    }
    Ok(())
}

/// Progress sink that emits the throttled `arbor://fs-op-progress` event through
/// the backend event sink. The totals arrive in `start`; per-file frames are
/// throttled to ~80ms, while the initial (0%) and final (100%) frames are forced
/// so the bar appears and completes. The payload (`op_id`/`kind`/`done_files`/
/// `total_files`/`done_bytes`/`total_bytes`/`current`) is byte-identical to the
/// old `AppHandle`-based emit.
struct EmitSink {
    sink:        Arc<dyn EventSink>,
    op_id:       String,
    kind:        &'static str,
    total_files: u64,
    total_bytes: u64,
    last_emit:   Instant,
}

impl EmitSink {
    fn emit(&mut self, done_files: u64, done_bytes: u64, current: &str, force: bool) {
        if !force && self.last_emit.elapsed().as_millis() < 80 {
            return;
        }
        self.last_emit = Instant::now();
        self.sink.emit("arbor://fs-op-progress", serde_json::json!({
            "op_id":       self.op_id,
            "kind":        self.kind,
            "done_files":  done_files,
            "total_files": self.total_files,
            "done_bytes":  done_bytes,
            "total_bytes": self.total_bytes,
            "current":     current,
        }));
    }
}

impl ProgressSink for EmitSink {
    fn start(&mut self, total_files: u64, total_bytes: u64) {
        self.total_files = total_files;
        self.total_bytes = total_bytes;
        self.emit(0, 0, "", true);
    }
    fn file_done(&mut self, done_files: u64, done_bytes: u64, current: &str) {
        self.emit(done_files, done_bytes, current, false);
    }
    fn finish(&mut self, done_files: u64, done_bytes: u64) {
        self.emit(done_files, done_bytes, "", true);
    }
}

/// Run an op closure with the right progress sink: an [`EmitSink`] when the
/// caller supplied an `op_id` (live progress), or a no-op sink otherwise.
fn run_op<F>(
    sink: &Arc<dyn EventSink>,
    op_id: &Option<String>,
    kind: &'static str,
    f: F,
) -> Result<Vec<String>, AppError>
where
    F: FnOnce(&mut dyn ProgressSink) -> arbor_fs::prelude::Result<Vec<String>>,
{
    let res = match op_id {
        Some(id) => {
            let mut emit = EmitSink {
                sink: Arc::clone(sink),
                op_id: id.clone(),
                kind,
                total_files: 0,
                total_bytes: 0,
                last_emit: Instant::now(),
            };
            f(&mut emit)
        }
        None => f(&mut NoopSink),
    };
    res.map_err(AppError::from)
}

/// Copy each of `sources` into `dest_dir`. With `overwrite = false` (default)
/// name collisions are resolved Explorer-style (" (2)", " (3)", …); with
/// `overwrite = true` each item keeps its name and merges into any existing
/// folder of the same name, replacing colliding files (recursive). Returns the
/// list of created / merged destination paths.
#[platform::handler(program = "platform")]
fn fs_copy(
    state: &AppState,
    sources: Vec<String>,
    dest_dir: String,
    overwrite: Option<bool>,
    op_id: Option<String>,
) -> Result<Vec<String>, AppError> {
    let sink = state
        .event_sink()
        .ok_or_else(|| AppError::Other("event sink unavailable".into()))?;
    let overwrite = overwrite.unwrap_or(false);
    let cancel = match &op_id {
        Some(id) => register_op(id),
        None => CancelToken::new(),
    };
    let out = run_op(&sink, &op_id, "copy", |sink| {
        copy::copy(&sources, &dest_dir, overwrite, sink, &cancel)
    });
    if let Some(id) = &op_id {
        unregister_op(id);
    }
    out
}

/// Move each of `sources` into `dest_dir` (cut + paste). Falls back to
/// copy-then-delete across volumes where `rename` can't work. With
/// `overwrite = true` an item keeps its name and merges into / replaces an
/// existing same-named entry instead of getting a " (2)" suffix. Returns the
/// list of new paths. Moving into the same directory is a no-op.
#[platform::handler(program = "platform")]
fn fs_move(
    state: &AppState,
    sources: Vec<String>,
    dest_dir: String,
    overwrite: Option<bool>,
    op_id: Option<String>,
) -> Result<Vec<String>, AppError> {
    let sink = state
        .event_sink()
        .ok_or_else(|| AppError::Other("event sink unavailable".into()))?;
    let overwrite = overwrite.unwrap_or(false);
    let cancel = match &op_id {
        Some(id) => register_op(id),
        None => CancelToken::new(),
    };
    let out = run_op(&sink, &op_id, "move", |sink| {
        copy::move_(&sources, &dest_dir, overwrite, sink, &cancel)
    });
    if let Some(id) = &op_id {
        unregister_op(id);
    }
    out
}

/// Duplicate each of `paths` in place (same parent folder), Explorer-style:
/// `report.pdf` → `report (2).pdf`, a second time → `report (3).pdf`. Returns
/// the created paths. Progress/cancel work exactly like `fs_copy`.
#[platform::handler(program = "platform")]
fn fs_duplicate(
    state: &AppState,
    paths: Vec<String>,
    op_id: Option<String>,
) -> Result<Vec<String>, AppError> {
    let sink = state
        .event_sink()
        .ok_or_else(|| AppError::Other("event sink unavailable".into()))?;
    let cancel = match &op_id {
        Some(id) => register_op(id),
        None => CancelToken::new(),
    };
    let out = run_op(&sink, &op_id, "duplicate", |sink| {
        copy::duplicate(&paths, sink, &cancel)
    });
    if let Some(id) = &op_id {
        unregister_op(id);
    }
    out
}
